connmgr.go

```diff
@@ -578,14 +620,61 @@ func (nn *cmNotifee) Disconnected(n network.Network, c network.Conn) {
 	atomic.AddInt32(&cm.connCount, -1)
 }
 
+// OpenedStream is called by notifiers to inform that a new libp2p stream has been opened on a connection.
+// The notifee updates the BasicConnMgr accordingly to update the number of streams we have open on a connection.
+// We then use this information when deciding which connections to trim.
+func (nn *cmNotifee) OpenedStream(_ network.Network, stream network.Stream) {
+	cm := nn.cm()
+
+	p := stream.Conn().RemotePeer()
+	s := cm.segments.get(p)
+	s.Lock()
+	defer s.Unlock()
+
+	cinf, ok := s.peers[p]
+	if !ok {
+		log.Error("received stream open notification for peer we are not tracking: ", p)
+		return
+	}
+
+	c := stream.Conn()
+	_, ok = cinf.conns[c]
+	if !ok {
+		log.Error("received stream open notification for conn we are not tracking: ", p)
+		return
+	}
+
+	cinf.conns[c].lastStreamOpen = time.Now()
+	cinf.conns[c].nStreams++
+}
+
+// ClosedStream is called by notifiers to inform that an existing libp2p stream has been closed.
+func (nn *cmNotifee) ClosedStream(_ network.Network, stream network.Stream) {
+	cm := nn.cm()
+
+	p := stream.Conn().RemotePeer()
+	s := cm.segments.get(p)
+	s.Lock()
+	defer s.Unlock()
+
+	cinf, ok := s.peers[p]
+	if !ok {
+		log.Error("received stream close notification for peer we are not tracking: ", p)
+		return
+	}
+
+	c := stream.Conn()
+	_, ok = cinf.conns[c]
+	if !ok {
+		log.Error("received stream close notification for conn we are not tracking: ", p)
+		return
+	}
+
+	cinf.conns[c].nStreams--
```

Comment:
clown shoes... you are doing the map lookup twice, and it's not free. Just capture it in a variable in the first lookup.

Reply:
done.